File: src/ios/LukuIDSDK/Tests/LukuIDSDKTests/EnvelopeTests.swift
// SPDX-License-Identifier: Apache-2.0
import XCTest
@testable import LukuIDSDK
import CryptoKit

final class EnvelopeTests: XCTestCase {

    private func getValidEnvelope() -> [String: Any] {
        let fileManager = FileManager.default
        var currentURL = URL(fileURLWithPath: #file)
        var sampleURL: URL?
        
        while currentURL.path != "/" {
            let candidate = currentURL.appendingPathComponent("samples/envelopes/dev/1.0.0/valid_envelope.json")
            if fileManager.fileExists(atPath: candidate.path) {
                sampleURL = candidate
                break
            }
            currentURL = currentURL.deletingLastPathComponent()
        }
        
        guard let url = sampleURL, let data = try? Data(contentsOf: url) else {
            fatalError("Could not locate valid_envelope.json")
        }
        
        return try! JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
    }

    func testVerifyEnvelopeValid() throws {
        let envelope = getValidEnvelope()
        let issues = LukuFile.verifyEnvelope(envelope: envelope, options: LukuVerifyOptions(allowUntrustedRoots: false, skipCertificateTemporalChecks: true, trustProfile: "dev"))
        XCTAssertTrue(issues.isEmpty, "Expected no issues, got \(issues)")
    }

    func testVerifyEnvelopeKeyMismatch() throws {
        var envelope = getValidEnvelope()
        
        let privateKey = Curve25519.Signing.PrivateKey()
        let publicKeyData = privateKey.publicKey.rawRepresentation
        let pubKeyBase64 = publicKeyData.base64EncodedString()
        
        var device = envelope["device"] as? [String: Any] ?? [:]
        device["public_key"] = pubKeyBase64
        envelope["device"] = device
        
        let canonical = envelope["canonical_string"] as? String ?? ""
        let signatureData = try privateKey.signature(for: canonical.data(using: .utf8)!)
        envelope["signature"] = signatureData.base64EncodedString()

        let issues = LukuFile.verifyEnvelope(envelope: envelope, options: LukuVerifyOptions(allowUntrustedRoots: false, skipCertificateTemporalChecks: true, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "ATTESTATION_FAILED" })
    }

    func testVerifyEnvelopeInvalidSignature() throws {
        var envelope = getValidEnvelope()
        envelope["signature"] = Data(repeating: 0, count: 64).base64EncodedString()

        let issues = LukuFile.verifyEnvelope(envelope: envelope, options: LukuVerifyOptions(allowUntrustedRoots: false, skipCertificateTemporalChecks: true, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "RECORD_SIGNATURE_INVALID" })
    }

    func testVerifyEnvelopeMissingIdentity() {
        var envelope = getValidEnvelope()
        envelope.removeValue(forKey: "device")
        envelope.removeValue(forKey: "device_id")
        envelope.removeValue(forKey: "public_key")
        
        let issues = LukuFile.verifyEnvelope(envelope: envelope, options: LukuVerifyOptions(allowUntrustedRoots: false, skipCertificateTemporalChecks: true, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "DEVICE_IDENTITY_MISSING" })
    }

    func testVerifyEnvelopeInvalidDac() throws {
        var envelope = getValidEnvelope()
        envelope["attestation_intermediate_der"] = "bad_cert".data(using: .utf8)?.base64EncodedString()
        let issues = LukuFile.verifyEnvelope(envelope: envelope, options: LukuVerifyOptions(allowUntrustedRoots: false, skipCertificateTemporalChecks: true, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "ATTESTATION_FAILED" })
    }

    func testVerifyEnvelopeInvalidCanonicalString() {
        var envelope = getValidEnvelope()
        envelope["canonical_string"] = "tampered:canonical:string"
        let issues = LukuFile.verifyEnvelope(envelope: envelope, options: LukuVerifyOptions(allowUntrustedRoots: false, skipCertificateTemporalChecks: true, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "RECORD_SIGNATURE_INVALID" })
    }

    func testVerifyEnvelopeTamperedMldsaSignature() throws {
        var envelope = getValidEnvelope()
        guard let derBase64 = envelope["attestation_intermediate_der"] as? String,
              var der = Data(base64Encoded: derBase64) else {
            XCTFail("Missing attestation_intermediate_der")
            return
        }
        // Tamper with the last byte of the signature
        der[der.count - 1] = der[der.count - 1] ^ 0xFF
        envelope["attestation_intermediate_der"] = der.base64EncodedString()
        
        let issues = LukuFile.verifyEnvelope(envelope: envelope, options: LukuVerifyOptions(allowUntrustedRoots: false, skipCertificateTemporalChecks: true, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "ATTESTATION_FAILED" }, "Expected ATTESTATION_FAILED due to tampered signature")
    }

    func testVerifyEnvelopeWrongRoot() throws {
        var envelope = getValidEnvelope()
        // Remove the intermediate so it tries to verify DAC against roots directly
        envelope.removeValue(forKey: "attestation_intermediate_der")
        if var identity = envelope["identity"] as? [String: Any] {
            identity.removeValue(forKey: "attestation_intermediate_der")
            envelope["identity"] = identity
        }
        
        let issues = LukuFile.verifyEnvelope(envelope: envelope, options: LukuVerifyOptions(allowUntrustedRoots: false, skipCertificateTemporalChecks: true, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "ATTESTATION_FAILED" }, "Expected ATTESTATION_FAILED due to missing/wrong root chain, got: \(issues)")
    }

    func testVerifyEnvelopeMalformedSpki() throws {
        var envelope = getValidEnvelope()
        guard let derBase64 = envelope["attestation_intermediate_der"] as? String,
              var der = Data(base64Encoded: derBase64) else {
            XCTFail("Missing attestation_intermediate_der")
            return
        }
        // Mess up the middle of the cert where SPKI usually is
        for i in 100..<200 { der[i] = 0x00 }
        envelope["attestation_intermediate_der"] = der.base64EncodedString()
        
        let issues = LukuFile.verifyEnvelope(envelope: envelope, options: LukuVerifyOptions(allowUntrustedRoots: false, skipCertificateTemporalChecks: true, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "ATTESTATION_FAILED" }, "Expected ATTESTATION_FAILED due to malformed SPKI")
    }

    func testVerifyEnvelopeMalformedTbs() throws {
        var envelope = getValidEnvelope()
        guard let derBase64 = envelope["attestation_intermediate_der"] as? String,
              var der = Data(base64Encoded: derBase64) else {
            XCTFail("Missing attestation_intermediate_der")
            return
        }
        // Mess up the start of the TBS SEQUENCE
        der[4] = 0xFF
        der[5] = 0xFF
        envelope["attestation_intermediate_der"] = der.base64EncodedString()
        
        let issues = LukuFile.verifyEnvelope(envelope: envelope, options: LukuVerifyOptions(allowUntrustedRoots: false, skipCertificateTemporalChecks: true, trustProfile: "dev"))
        XCTAssertTrue(issues.contains { $0.code == "ATTESTATION_FAILED" }, "Expected ATTESTATION_FAILED due to malformed TBS")
    }
}

File: src/rust/lukuid-sdk/tests/envelope_tests.rs
use lukuid_sdk::{LukuFile, LukuVerifyOptions};
use serde_json::Value;
use std::fs;
use std::path::{Path, PathBuf};
use ed25519_dalek::{SigningKey, Signer};
use base64::{Engine as _, engine::general_purpose::STANDARD as BASE64};

fn resolve_sample() -> PathBuf {
    let mut current = std::env::current_dir().unwrap();
    loop {
        let candidate = current.join("samples").join("envelopes").join("dev").join("1.0.0").join("valid_envelope.json");
        if candidate.exists() {
            return candidate;
        }
        if !current.pop() {
            break;
        }
    }
    PathBuf::from("../../../samples/envelopes/dev/1.0.0/valid_envelope.json")
}

fn get_valid_envelope() -> Value {
    let path = resolve_sample();
    let content = fs::read_to_string(&path).expect("Failed to read valid_envelope.json");
    serde_json::from_str(&content).expect("Failed to parse valid_envelope.json")
}

#[test]
fn test_verify_envelope_valid() {
    let envelope = get_valid_envelope();
    let mut options = LukuVerifyOptions::default();
    options.allow_untrusted_roots = false;
    options.skip_certificate_temporal_checks = true;
    options.trust_profile = "dev".to_string();

    let issues = LukuFile::verify_envelope(&envelope, options);
    assert!(issues.is_empty(), "Expected no issues, got: {:?}", issues);
}

use rand_core::OsRng;

#[test]
fn test_verify_envelope_key_mismatch() {
    let mut envelope = get_valid_envelope();
    
    let mut csprng = OsRng;
    let signing_key = SigningKey::generate(&mut csprng);
    let verifying_key = signing_key.verifying_key();
    let new_pub_key_base64 = BASE64.encode(verifying_key.as_bytes());

    if let Some(device) = envelope.get_mut("device").and_then(|d| d.as_object_mut()) {
        device.insert("public_key".to_string(), Value::String(new_pub_key_base64));
    }

    let canonical = envelope["canonical_string"].as_str().unwrap();
    let signature = signing_key.sign(canonical.as_bytes());
    envelope["signature"] = Value::String(BASE64.encode(signature.to_bytes()));

    let mut options = LukuVerifyOptions::default();
    options.allow_untrusted_roots = false;
    options.skip_certificate_temporal_checks = true;
    options.trust_profile = "dev".to_string();

    let issues = LukuFile::verify_envelope(&envelope, options);
    assert!(issues.iter().any(|i| i.code == "ATTESTATION_FAILED"), "Expected ATTESTATION_FAILED");
}

#[test]
fn test_verify_envelope_invalid_signature() {
    let mut envelope = get_valid_envelope();
    envelope["signature"] = Value::String(BASE64.encode(vec![0u8; 64]));

    let mut options = LukuVerifyOptions::default();
    options.allow_untrusted_roots = false;
    options.skip_certificate_temporal_checks = true;
    options.trust_profile = "dev".to_string();

    let issues = LukuFile::verify_envelope(&envelope, options);
    assert!(issues.iter().any(|i| i.code == "RECORD_SIGNATURE_INVALID"), "Expected RECORD_SIGNATURE_INVALID");
}

#[test]
fn test_verify_envelope_missing_identity() {
    let mut envelope = get_valid_envelope();
    if let Some(obj) = envelope.as_object_mut() {
        obj.remove("device");
        obj.remove("device_id");
        obj.remove("public_key");
    }

    let mut options = LukuVerifyOptions::default();
    options.allow_untrusted_roots = false;
    options.skip_certificate_temporal_checks = true;
    options.trust_profile = "dev".to_string();

    let issues = LukuFile::verify_envelope(&envelope, options);
    assert!(issues.iter().any(|i| i.code == "DEVICE_IDENTITY_MISSING"), "Expected DEVICE_IDENTITY_MISSING");
}

#[test]
fn test_verify_envelope_invalid_dac() {
    let mut envelope = get_valid_envelope();
    envelope["attestation_intermediate_der"] = Value::String(BASE64.encode(b"bad_cert"));

    let mut options = LukuVerifyOptions::default();
    options.allow_untrusted_roots = false;
    options.skip_certificate_temporal_checks = true;
    options.trust_profile = "dev".to_string();

    let issues = LukuFile::verify_envelope(&envelope, options);
    assert!(issues.iter().any(|i| i.code == "ATTESTATION_FAILED"), "Expected ATTESTATION_FAILED");
}

#[test]
fn test_verify_envelope_invalid_canonical_string() {
    let mut envelope = get_valid_envelope();
    envelope["canonical_string"] = Value::String("tampered:canonical:string".to_string());

    let mut options = LukuVerifyOptions::default();
    options.allow_untrusted_roots = false;
    options.skip_certificate_temporal_checks = true;
    options.trust_profile = "dev".to_string();

    let issues = LukuFile::verify_envelope(&envelope, options);
    assert!(issues.iter().any(|i| i.code == "RECORD_SIGNATURE_INVALID"), "Expected RECORD_SIGNATURE_INVALID");
}
